### DISCUSSION TITLE
RT10XX low frequency PWM with external clock

### DISCUSSION THREAD

**Discussion by bbilas** [zephyrproject-rtos/zephyr#51006] [Q&A] [answered]
Hello,
I'm trying to configure PWM with a low-frequency output (10Hz) using RT1062 MCU. I'm unable to do that out-of-the-box since the PWM clock is fed from IPG bus which is 150MHz and with 128 prescaler the period does not fit into the 16-bit register. Since IPG bus clock is related to many other peripherals I don't want to change it, that is why I want to use the external clock for PWM that will achieve the intended purpose.

![image](https://user-images.githubusercontent.com/8523351/194119909-f0ce910f-01ec-44c8-be12-eb23272cc980.png)

![image](https://user-images.githubusercontent.com/8523351/194121975-c91aa2da-ddcd-49bb-9641-b0779ed9e6cb.png)

Regarding NXP forum information [1], the external PWM clock can reach the half of IPG clock which is 75MHz which would be fair enough for me. To use it there is a necessity to configure an XBAR peripheral that allows feeding the PWM clock from that but unfortunately, there is no example or information on how to use it properly. 

[1] <pre>I would like to apologize for the delay. We are overloaded on the request these days. I really really appreciate that you were so patient.

 

Regarding your question about the EXT_CLK for FlexPWM2, after double-checking internally, this external clock only can reach to the half of internal ipg_clock, that its default is 150MHz, so the maximum frequency you can drive on FLEXPWMn_EXT_CLK is 75MHz. Sadly, It is not stated in any public documentation.

 

Please let me know if you have more questions.

 

Best Regards.
Pablo Avalos.</pre>

I've created a topic in the Zephyr discord NXP channel but I was told to create a new discussion here so I would be appreciated for any hints.

**Comment by DerekSnell** [comment, marked as answer]
Hi @bbilas ,
Attached is a simple example project to help you [evkbmimxrt1060_qtmr_xbar_flexpwm_v2.zip](https://github.com/zephyrproject-rtos/zephyr/files/9724544/evkbmimxrt1060_qtmr_xbar_flexpwm_v2.zip).  As discussed, this is an MCUXpresso SDK-based project, using MCUXpresso IDE.

The details below are from the included readme.  I hope this helps you configure the hardware for your use-case.  Best regards

Overview
========
This example demonstrates driving a FlexPWM output at a very low frequency of 10Hz, 
using the EXT_CLK source driven by a QTMR PWM output, connected through the XBAR.

Toolchain supported
===================
- MCUXpresso  11.6.0, with SDK_2.12.1_MIMXRT1060-EVKB

Hardware requirements
=====================
- Mini/micro USB cable
- MIMXRT1060-EVKB board
- Personal Computer
- Oscilloscope

Board settings
==============

* Remove J4, disconnects JTAG_TDI from GPIO_AD_B0_09

* Probe the 10Hz PWM signal using an oscilloscope
 - At J16-5 (signal FLEXPWM2_PWMA03 on GPIO_AD_B0_09)

* optionally, can probe the 600kHz clock signal from QTMR4 to FlexPWM2, driven to output pin by XBAR
 - At J17-2 (signal XBAR1_INOUT16 on GPIO_AD_B0_02)

Prepare the Demo
================
1. Connect a USB cable between the PC host and the OpenSDA USB port on the board.
2. Download the program to the target board.
3. Either press the reset button on your board or launch the debugger in your IDE to begin running the example.

**Comment by bbilas** [comment]
~Hello @DerekSnell,
many thanks for your answer! I've tried to use your example but it does not work well, I mean - I see that 600kHz which is generated by QTMR4 but there is no 10Hz PWM output on the GPIO_AD_B0_09 pin (I guess there is also a typo in the readme, this signal is not available on the J4 socket but J16). Would you mind double-checking that?~

I've commented too many lines that's why it didn't work well. Sorry for the noise!

**Reply by DerekSnell** [reply]
Hi @bbilas ,
I am glad it works for you now.  You raise a good point, the 10Hz PWM on GPIO_AD_B0_09 is available on J16-5.  It is also available on J4 as JTAG_TDI.  But I like using J16-5 better, and J4 can be removed if the JTAG_TDI signal interferes with the PWM output.  I have updated the readme in the example project, and in my comment above.

Best regards

**Comment by bbilas** [comment]
FYI: I have successfully integrated that with the existing PWM driver in my downstream project where a 1-1kHz PWM frequencies range was required. Many thanks, @DerekSnell.
